src/control/cmd/daos/filesystem.go

```diff
@@ -69,73 +67,217 @@ func (cmd *fsCopyCmd) Execute(_ []string) error {
 type fsAttrCmd struct {
 	existingContainerCmd
 
-	ChunkSize   ChunkSizeFlag `long:"chunk-size" short:"z" description:"container chunk size"`
-	ObjectClass ObjClassFlag  `long:"oclass" short:"o" description:"default object class"`
-	DfsPath     string        `long:"dfs-path" short:"H" description:"DFS path relative to root of container, when using pool and container instead of --path and the UNS"`
-	DfsPrefix   string        `long:"dfs-prefix" short:"I" description:"Optional prefix path to the root of the DFS container when using pool and container"`
+	DfsPath   string `long:"dfs-path" short:"H" description:"DFS path relative to root of container, when using pool and container instead of --path and the UNS"`
+	DfsPrefix string `long:"dfs-prefix" short:"I" description:"Optional prefix path to the root of the DFS container when using pool and container"`
 }
 
-func (cmd *fsAttrCmd) Execute(_ []string) error {
-	ap, deallocCmdArgs, err := allocCmdArgs(cmd.log)
-	if err != nil {
-		return err
-	}
-	defer deallocCmdArgs()
-
+func parseDFSParamsToAP(cmd *fsAttrCmd, ap *C.struct_cmd_args_s) error {
 	if cmd.DfsPath != "" {
 		if cmd.Path != "" {
-			return errors.New("Cannot use both --dfs-path and --path")
+			return errors.New("cannot use both --dfs-path and --path")
 		}
 		ap.dfs_path = C.CString(cmd.DfsPath)
 	}
 	if cmd.DfsPrefix != "" {
 		if cmd.Path != "" {
-			return errors.New("Cannot use both --dfs-prefix and --path")
+			return errors.New("cannot use both --dfs-prefix and --path")
 		}
 		ap.dfs_prefix = C.CString(cmd.DfsPrefix)
 	}
+	return nil
+}
 
-	flags := C.uint(C.DAOS_COO_RW)
-	op := os.Args[2]
-	switch op {
-	case "set-attr":
-		ap.fs_op = C.FS_SET_ATTR
-		if cmd.ObjectClass.Set {
-			ap.oclass = cmd.ObjectClass.Class
+func setupFSAttrCmd(cmd *fsAttrCmd, flags C.uint, op uint32) (*C.struct_cmd_args_s, *C.dfs_t, func(), error) {
```

Comment:
most of this is already done in the C handler.. why are we converting it to go now?

Reply:
The problem was that the C handler wasn't able to print JSON output for get-attr. The easiest way to resolve it was moving the logic into Go. Since so much logic is shared, it seemed better to move the whole group of commands instead of just get-attr. One nice effect is that the errors returned in the JSON results for all of these commands are more descriptive now, since we can use the error strings that were previously dumped only to stderr.

Reply:
We can't use like json-c to print the output in the C handler?

My concern is that C is what many devs in our group know. golang on the other hand is not.
so moving more logic here is just downsizing how many people can properly fix bugs and update this code.
we have also seen issues with golang and valgrind with unexplained memory leaks. see: https://daosio.atlassian.net/browse/DAOS-8404
so im worried doing more of this might be increasing those issues.
so im not in favor of this TBH, but if this is a restructuring we can discuss or already discussed and agreed on, then i won't object ..

Reply:
I appreciate your concerns. I was a little worried about switching such a big chunk of code to Go, but it seemed like the easiest path. I don't think using json-c is necessarily feasible since it creates a second way to do the same thing we are doing automatically (or nearly so) with Go, and seems likely to make things a little more confusing for that reason. I could try another approach, which would be breaking out the get-attr C code to let it return a struct, instead of printing to stdout. Then the Go code could decide on the output formatting. That would keep most of the logic in C.

An aside, my PR (plus a fix I haven't pushed yet due to the CI issues) doesn't reproduce the sporadic leak using your reproducer code in DAOS-8404.